sdk/core/src/main/java/com/klaviyo/coresdk/networking/requests/KlaviyoApiRequest.kt

```diff
@@ -0,0 +1,234 @@
+                query = json.getJSONObject(QUERY_JSON_KEY).let {
+                    it.keys().asSequence().associateWith { k -> it.getString(k) }
+                }
+                body = json.optJSONObject(BODY_JSON_KEY)
+            }
+        }
+
+        /**
+         * Compare two requests purely by their UUIDs
+         *
+         * @param request1
+         * @param request2
+         * @return
+         */
+        fun equals(request1: KlaviyoApiRequest, request2: KlaviyoApiRequest): Boolean =
+            request1.uuid == request2.uuid
+    }
+
+    /**
+     * Compiles the base url, path and query data into a [URL] object
+     */
+    val url: URL
+        get() {
+            val baseUrl = Registry.config.baseUrl
+            val queryMap = query.map { (key, value) -> "$key=$value" }
+            val queryString = queryMap.joinToString(separator = "&")
+
+            return if (queryString.isEmpty())
+                URL("$baseUrl/$urlPath")
+            else
+                URL("$baseUrl/$urlPath?$queryString")
+        }
+
+    /**
+     * Builds and sends a network request to Klaviyo and then parses and handles the response
+     *
+     * @returns The string value of the response body, if one was returned
+     */
+    fun send(): String? {
+        if (!Registry.networkMonitor.isNetworkConnected()) {
+            return null
+        }
+
+        val connection = buildUrlConnection()
+
+        return try {
+            connection.connect()
+            return readResponse(connection)
+        } catch (ex: IOException) {
+            // TODO Logging - TP-164154
+            return null
+        } finally {
+            connection.disconnect()
+        }
+    }
+
+    /**
+     * Opens a connection against the given [URL]
+     * Connection type is either [HttpURLConnection] or [HttpsURLConnection] depending on the [URL] protocol
+     */
+    private fun buildUrlConnection(): HttpURLConnection {
+        val connection = HttpUtil.openConnection(url)
+
+        headers.forEach { (key, header) ->
+            connection.setRequestProperty(key, header)
+        }
+
+        connection.requestMethod = method.name
+        connection.readTimeout = Registry.config.networkTimeout
+        connection.connectTimeout = Registry.config.networkTimeout
+
+        val bodyString = body?.toString() ?: return connection
+
+        connection.doOutput = true
+        HttpUtil.writeToConnection(bodyString, connection)
+
+        return connection
+    }
+
+    /**
+     * Reads the response body from the open [HttpURLConnection]
+     * If the request was successful, extracts the response body
+     * If the request was unsuccessful, extracts the error response body
+     *
+     * @return The string extracted from the response body
+     */
+    private fun readResponse(connection: HttpURLConnection): String {
+        val statusCode = connection.responseCode
+        val stream = if (statusCode in HTTP_OK until HTTP_MULT_CHOICE) {
```

Comment:
does this perform well? or will it create a list from OK to MULT_CHOICE and iterate until statusCode matches an element in the list?

Reply:
can we do a comparison check instead?

Reply:
Hum I just copied this, I'll take a closer look

Reply:
Yeah i think you're right this is basically like saying `if (200 in [1,2,3,4,5,6,7...])` using greater than/less than makes more sense. Syntactic sugar is so tempting.

Reply:
Huh https://stackoverflow.com/questions/69171532/is-kotlins-range-checking-is-efficent 
The IDE tells me to use range check too... and who am I to disobey android studio

Reply:
oooo very nice